src/EVMBlackjack.sol

```diff
@@ -0,0 +1,337 @@
+        game.playerHands.push(Hand({cards: new uint8[](0), betSize: _betSize}));
+
+        // Request randomness
+        requestRandomness(msg.sender);
+
+        emit BetPlaced(msg.sender, _betSize);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+    //  Insurance
+    //////////////////////////////////////////////////////////////*/
+
+    function takeInsurance(bool _take) public {
+        Game storage game = games[msg.sender];
+
+        if (game.state != State.READY_FOR_INSURANCE) {
+            revert InvalidStateTransition();
+        }
+
+        if (_take) {
+            uint16 insuranceBet = game.playerHands[0].betSize / 2;
+
+            // Store side bet
+            game.insurance = insuranceBet;
+
+            // Transfer CHIP
+            chip.transferFrom(msg.sender, address(this), insuranceBet);
+        }
+
+        // Update game state
+        game.state = State.READY_FOR_PLAYER_ACTION;
+
+        emit InsuranceTaken(msg.sender, _take);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+    //  Player Action
+    //////////////////////////////////////////////////////////////*/
+
+    function takeAction(Action action) public {
+        Game storage game = games[msg.sender];
+
+        if (game.state != State.READY_FOR_PLAYER_ACTION) {
+            revert InvalidStateTransition();
+        }
+
+        // Handle player action.
+        if (action == Action.SPLIT_ACES) {
+            if (!isAce(game.playerHands[0].cards[0]) || !isAce(game.playerHands[0].cards[1])) {
+                revert InvalidStateTransition();
+            }
+
+            // TODO instantiate new hand and request randomness
+
+            game.state = State.WAITING_FOR_RANDOMNESS;
+            game.lastAction = Action.SPLIT_ACES;
+        } else if (action == Action.SPLIT) {
+            if (!isPair(game.playerHands[0].cards[0], game.playerHands[0].cards[1])) {
+                revert InvalidStateTransition();
+            }
+
+            // TODO instantiate new hand and request randomness
+
+            game.state = State.WAITING_FOR_RANDOMNESS;
+            game.lastAction = Action.SPLIT;
+        } else if (action == Action.DOUBLE_DOWN) {
+            // TODO request randomness
+
+            game.state = State.WAITING_FOR_RANDOMNESS;
+            game.lastAction = Action.DOUBLE_DOWN;
+            uint16 betSize = game.playerHands[0].betSize;
+            game.playerHands[0].betSize *= 2;
+
+            // Transfer CHIP
+            chip.transferFrom(msg.sender, address(this), betSize);
+
+            // QUESTION should we emit an event like BetIncreased ?
+        } else if (action == Action.HIT) {
+            // TODO request randomness
+
+            game.state = State.WAITING_FOR_RANDOMNESS;
+            game.lastAction = Action.HIT;
+        } else if (action == Action.STAND) {
+            game.state = State.DEALER_ACTION;
```

Comment:
i think every state in here should transition to waiting for randomness, no?

Reply:
here, we request randomness for the dealer.

Reply:
Probably you are right. Need to pull on the dealer action thread a bit more but I get what you mean.

Reply:
Yeah, it'll still need to request/await randomness here @neodaoist